### PR TITLE
ROX-23346: crd-schema-checker should check for removed enums

### PR DESCRIPTION
**Pull request by JoukoVirtanen**
Currently crd-schema-checker does not check for removed enums. This PR fixes that situation.

### CODE REVIEW

File: pkg/manifestcomparators/comp_no_field_removal.go [hunk 1]
```diff
@@ -43,16 +43,32 @@ func (b noFieldRemoval) Compare(existingCRD, newCRD *apiextensionsv1.CustomResou
...
 		SchemaHas(newVersion.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
 			func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
 				newFields.Insert(simpleLocation.String())
+				for _, enum := range s.Enum {
+					_, exists := newEnumsMap[simpleLocation.String()]
+					if !exists {
+						newEnumsMap[simpleLocation.String()] = sets.NewString()
+					}
+					newEnumsMap[simpleLocation.String()].Insert(string(enum.Raw))
+				}
```
Comment (JoukoVirtanen) [inline pkg/manifestcomparators/comp_no_field_removal.go:71]: This needs to be refactored so that there is no repetition.

**Comment by JoukoVirtanen** [comment]
I added the changes to `pkg/manifestcomparators/comp_no_field_removal.go` this seems to make more sense than putting the check for removed enums into a separate file, since it is not possible to check for missing enums without iterating through the fields, and missing fields can be reported at the same time. However, there is a case that these should be separate checks in separate files. If the changes are kept in this file, the file should be renamed to reflect the fact that it is also checking for removed enums.

File: pkg/manifestcomparators/comp_no_field_removal.go (hunk 1 above)
Comment (JoukoVirtanen) [reply pkg/manifestcomparators/comp_no_field_removal.go]: Done

File: pkg/manifestcomparators/comp_no_field_removal.go [hunk 2]
```diff
@@ -22,6 +22,26 @@ func (noFieldRemoval) WhyItMatters() string {
 	return "If fields are removed, then clients that rely on those fields will not be able to read them or write them."
 }
 
+func getFieldsAndEnums(version *apiextensionsv1.CustomResourceDefinitionVersion) (sets.String, map[string]sets.String) {
+	fields := sets.NewString()
+	enumsMap := make(map[string]sets.String)
+	SchemaHas(version.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+		func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
+			fields.Insert(simpleLocation.String())
+			for _, enum := range s.Enum {
```
Comment (deads2k) [inline pkg/manifestcomparators/comp_no_field_removal.go:31]: This is a validation rule, not a new location right?  I'm in favor of the rule, I'd prefer a new file to check for "no_enumeration_removal" versus checking both fields and validation from one checker

**Comment by deads2k** [comment]
definitely in favor of this update.

File: pkg/manifestcomparators/comp_no_field_removal.go (hunk 2 above)
Comment (JoukoVirtanen) [reply pkg/manifestcomparators/comp_no_field_removal.go]: I moved the check to a new file.

File: pkg/manifestcomparators/comp_no_field_removal.go [hunk 3]
```diff
@@ -22,6 +22,18 @@ func (noFieldRemoval) WhyItMatters() string {
...
+	fields := sets.NewString()
+	SchemaHas(version.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+		func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
+			fields.Insert(simpleLocation.String())
+			return false
+		})
+
+	return fields
+}
+
```
Comment (JoukoVirtanen) [inline pkg/manifestcomparators/comp_no_field_removal.go:36]: Not needed for this PR, but it seems like a good refactor.

**Comment by deads2k** [comment]
looks good.  Can I get a squash and a test that ensures when a new version is created with fewer enum values in the "same" field there is no error?

Also, wouldn't mind having the extraneous bits of the test schemas removed.

File: pkg/manifestcomparators/comp_no_field_removal.go (hunk 3 above)
Comment (deads2k) [reply pkg/manifestcomparators/comp_no_field_removal.go:35]: Agreed

**Comment by JoukoVirtanen** [comment]
> looks good. Can I get a squash and a test that ensures when a new version is created with fewer enum values in the "same" field there is no error?
> 
> Also, wouldn't mind having the extraneous bits of the test schemas removed.

I am not sure why you put "same" in quotation marks. Are you asking for a test that has multiple fields with enums?

**Pull request merged**